tests/test_run_config_generator.py

```diff
@@ -71,6 +72,286 @@ def test_default_config_single_model(self):
+                        max_batch_size: [1,4,16,64]
+                        instance_group:
+                        -
+                            kind: KIND_GPU
+                            count: [1,2,3]
+            """)
+        # yapf: enable
+
+        expected_num_of_configs = 9750
+
+        # Expect 110 calls to ModelRunConfigGenerator.set_last_results
+        # All 9750 experiments will be passed to the leaf generator
+        # All 250 times that the leaf generator is done will pass results to the middle generator
+        # All 10 times that the middle generator is done will pass results to the root generator
+        expected_num_calls_to_set_last_results = 10010
+
+        with patch.object(ModelRunConfigGenerator,
+                          "set_last_results") as mock_method:
+            self._run_and_test_run_config_generator(
+                yaml_content, expected_config_count=expected_num_of_configs)
+
+            self.assertEqual(mock_method.call_count,
+                             expected_num_calls_to_set_last_results)
+
+    def test_early_backoff_leaf_model(self):
+        """
+        Test the case where there are two models, and the 'leaf' model has early backoff
+        
+        Both models are auto search:
+            num_PAC = 2
+            num_MC = (4 * 2) + 1 = 9
+            model_total = (2 * 5) = 18
+
+        total = model1_total * model2_total = 324
+
+        However, the test will set up the throughput values such that Model2 will see a lack 
+        of throughput the first time that it is walking max_batch_size. 
+        Normally it would walk values 1,2,4,8, but for this ONE case it will only walk 1,2.
+        This will reduce the total count by 4, as there are two concurrencies that would be 
+        tried for Model2 for each of the two removed max_batch_sizes
+
+        Thus, actual expected_count = 320
+
+
+        """
+
+        # yapf: disable
+        yaml_content = convert_to_bytes("""
+            run_config_search_max_model_batch_size: 8
+            run_config_search_max_instance_count: 2
+            run_config_search_max_concurrency: 2
+            profile_models:
+                - my-model
+                - my-model2
+            """)
+        # yapf: enable
+
+        expected_num_of_configs = 320
+
+        perf_throughput_values = [2**i for i in range(expected_num_of_configs)]
+        perf_throughput_values[4] = perf_throughput_values[3]
+        perf_throughput_values[5] = perf_throughput_values[3]
+
+        with patch.object(TestRunConfigGenerator,
+                          "_get_next_perf_throughput_value") as mock_method:
+            mock_method.side_effect = perf_throughput_values
+            self._run_and_test_run_config_generator(
+                yaml_content, expected_config_count=expected_num_of_configs)
+
+    def test_early_backoff_root_model(self):
+        """
+        Test the case where there are two models, and the 'root' model has early backoff
```

Comment:
By root do you mean model1? Maybe say first model instead

Reply:
I meant the first one called in the recursive generator stack. I've tried to clarify the comments